ash/src/lib.rs

```diff
@@ -66,6 +66,40 @@ impl<'r, T> RawPtr<T> for Option<&'r T> {
     }
 }
 
+/// Given a mutable raw pointer to a type with an `s_type` member such as [`vk::BaseOutStructure`],
+/// match on a set of Vulkan structures. The struct will be rebound to the
+/// given variable of the type of the given Vulkan structure.
+///
+/// ```
+/// let mut info = ash::vk::DeviceCreateInfo::default();
+/// let info: *mut ash::vk::BaseOutStructure = <*mut _>::cast(&mut info);
+/// unsafe {
+///     ash::match_struct!(match info {
+///         info @ ash::vk::DeviceQueueCreateInfo => {
+///             dbg!(&info); // Unreachable
+///         }
+///         info @ ash::vk::DeviceCreateInfo => {
+///             dbg!(&info);
+///         }
+///     })
+/// }
+/// ```
+#[macro_export]
+macro_rules! match_struct {
+    (match $p:ident { $($bind:ident @ $ty:path => $body:block $(,)?)+ }) => {
+        match (*$p).s_type {
```

Comment:
This derefs the entire `$p`, which might be unsound if any part of it is uninitialized, which could reasonably happen for out structs. Instead, we could use `ptr::addr_of` to compute the address of the field without dereffing, and then deref that field alone.

Reply:
@Ralith Will it temporarily copy all fields during dereferencing? I've previously seen a similar `&*` reborrowing technique being used and expect the `&` to be redundant here, but that probably makes no difference for the soundness issue here.

I however also remember previously having trouble with this method for converting from `*const` to `&` with `&*`, which was solved by simply transmuting. Leading to think that the compiler would perhaps run the destructor on the temporary created by `*` but that's not the case (and fwiw irrelevant on our Vulkan types since it's not allowed to implement both `Drop` and `Copy` simultaneously).

I should read up more on raw pointers :sweat_smile:

Reply:
I landed on `std::ptr::addr_of!((*$p).s_type).read_unaligned()` here (expression still needs to be valid), though all Vulkan structs should probably be aligned.

Reply:
> Will it temporarily copy all fields during dereferencing?

That's a fair mental model, though in formal terms I think it's more correct to say that it is the deref operation itself (which produces a place, not necessarily copying anything) that is UB.

> I however also remember previously having trouble with this method for converting from *const to & with &*, which was solved by simply transmuting

I'm not totally sure what you're alluding to here, but constructing a reference to uninitialized data is also UB.

Reply:
> That's a fair mental model, though in formal terms I think it's more correct to say that it is the deref operation itself (which produces a place, not necessarily copying anything) that is UB.

Wasn't considering copying to be a mnemonic to this, just curious if it is the case at all. Fair enough, retrieving a borrow to possibly uninitialized memory - even if unused otherwise - is UB. 

> > I however also remember previously having trouble with this method for converting from _const to & with &_, which was solved by simply transmuting
> 
> I'm not totally sure what you're alluding to here, but constructing a reference to uninitialized data is also UB.

I'll probably have to repro this case again, but it was happening on initialized data.

Reply:
And yeah, `read_unaligned` is overkill, but that use of `addr_of` looks fine.